File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/BreadthFirstSearch/102_BinaryTreeLevelOrderTraversal.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-24 10:59:23
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

c.. Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        __(root__NULL){
            r_ {};
        }
        vector<vector<int>> ans;
        deque<TreeNode*> levels{root};
        _____(!levels.empty()){
            vector<int> cur_level;
            int l_size = levels.size();
            ___(int i=0;i<l_size;i++){
                TreeNode* cur_node = levels.front();
                levels.pop_front();
                cur_level.push_back(cur_node->val);
                __(cur_node->left)  levels.push_back(cur_node->left);
                __(cur_node->right) levels.push_back(cur_node->right);
            }
            ans.push_back(cur_level);
        }
        r_ ans;
    }
};


/*
[]
[1]
[1,2,3]
[3,9,20,null,null,15,7]
*/

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/BreadthFirstSearch/102_BinaryTreeLevelOrderTraversal.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

c.. Solution o..
    ___ levelOrder  root
        __ n.. root:
            r_ []

        cur_level, ans = [root], []

        # Breadth-first Search, Pythonic way.
        _____ cur_level:
            ans.a.. [node.val ___ node __ cur_level])
            cur_level = [kid ___ n __ cur_level
                         ___ kid __ (n.left, n.right) __ kid]

        r_ ans

"""
[]
[1]
[1,2,3]
[3,9,20,null,null,15,7]
"""
